data/real/basic.lean

```diff
@@ -413,6 +417,66 @@ theorem equiv_lim (f : cau_seq ℝ abs) : f ≈ const abs (lim f) :=
 by simp [lim, f.is_cau]; cases f with f hf;
    exact classical.some_spec (cau_seq_converges ⟨f, hf⟩)
 
+lemma eq_lim_of_const_equiv {f : cau_seq ℝ abs} {x : ℝ} (h : cau_seq.const abs x ≈ f) : x = lim f :=
+const_equiv.mp (setoid.trans h (equiv_lim f))
+
+lemma lim_eq_of_equiv_const {f : cau_seq ℝ abs} {x : ℝ} (h : f ≈ cau_seq.const abs x) : lim f = x :=
+(eq_lim_of_const_equiv (setoid.symm h)).symm
+
+lemma lim_eq_lim_of_equiv {f g : cau_seq ℝ abs} (h : f ≈ g) : lim f = lim g := 
+lim_eq_of_equiv_const (setoid.trans h (equiv_lim g))
+
+@[simp] lemma lim_const (x : ℝ) : lim (const abs x) = x := 
+lim_eq_of_equiv_const (setoid.refl _)
+
+lemma lim_add (f g : cau_seq ℝ abs) : lim f + lim g = lim ⇑(f + g) := 
+eq_lim_of_const_equiv (show lim_zero (const abs (lim ⇑f + lim ⇑g) - (f + g)),
+  from by rw [const_add, add_sub_comm];
+  exact add_lim_zero (setoid.symm (equiv_lim f)) (setoid.symm (equiv_lim g)))
+
+lemma lim_mul_lim (f g : cau_seq ℝ abs) : lim f * lim g = lim ⇑(f * g) := 
+eq_lim_of_const_equiv (show lim_zero (const abs (lim ⇑f * lim ⇑g) - f * g),
+  from have h : const abs (lim ⇑f * lim ⇑g) - f * g = g * (const abs (lim f) - f) 
+      + const abs (lim f) * (const abs (lim g) - g) := 
+    by simp [mul_sub, mul_comm, const_mul, mul_add],
+  by rw h; exact add_lim_zero (mul_lim_zero _ (setoid.symm (equiv_lim f))) 
+      (mul_lim_zero _ (setoid.symm (equiv_lim g))))
+
+lemma lim_mul (f : cau_seq ℝ abs) (x : ℝ) : lim f * x = lim ⇑(f * const abs x) :=
+by rw [← lim_mul_lim, lim_const]
+
+lemma lim_neg (f : cau_seq ℝ abs) : lim ⇑(-f) = -lim f :=
+lim_eq_of_equiv_const (show lim_zero (-f - const abs (-lim ⇑f)),
+  from by rw [const_neg, sub_neg_eq_add, add_comm];
```

Comment:
`from by` is can be just written as `by`.

Reply:
done